libflux/src/libstd/build.rs

```diff
@@ -1,24 +1,62 @@
-use std::env;
-use std::fs::File;
-use std::io::Write;
-use std::path::PathBuf;
+use std::{env, fs, io, io::Write, path};
 
-fn prelude() -> Vec<u8> {
-    Vec::new()
+use flux::semantic::bootstrap;
+use flux::semantic::env::Environment;
+use flux::semantic::flatbuffers::types as fb;
+
+use flatbuffers;
+
+#[derive(Debug)]
+struct Error {
+    msg: String,
+}
+
+impl From<env::VarError> for Error {
+    fn from(err: env::VarError) -> Error {
+        Error {
+            msg: err.to_string(),
+        }
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Error {
+        Error {
+            msg: format!("{:?}", err),
+        }
+    }
 }
 
-fn stdlib() -> Vec<u8> {
-    Vec::new()
+impl From<bootstrap::Error> for Error {
+    fn from(err: bootstrap::Error) -> Error {
+        Error { msg: err.msg }
+    }
 }
 
-fn main() {
-    let dir = PathBuf::from(env::var("OUT_DIR").unwrap());
+fn serialize<'a, T, S, F>(ty: T, f: F, path: &path::Path) -> Result<(), Error>
+where
+    F: Fn(&mut flatbuffers::FlatBufferBuilder<'a>, T) -> flatbuffers::WIPOffset<S>,
+{
+    let mut builder = flatbuffers::FlatBufferBuilder::new();
+    let buf = fb::serialize(&mut builder, ty, f);
+    let mut file = fs::File::create(path)?;
+    file.write_all(&buf)?;
+    Ok(())
+}
+
+fn main() -> Result<(), Error> {
+    let dir = path::PathBuf::from(env::var("OUT_DIR")?);
+
+    let (pre, lib, fresher) = bootstrap::infer_stdlib()?;
+
+    let path = dir.join("prelude.data");
+    serialize(Environment::from(pre), fb::build_env, &path)?;
+
+    let path = dir.join("stdlib.data");
+    serialize(Environment::from(lib), fb::build_env, &path)?;
```

Comment:
`stdlib.data`, `prelude.data` and `fresher.data` are all generated files.  We typically check generated files into source control.  I guess we can't generate them as part of the `generate` step?

Do you need to add a `.gitignore` file for these files, if they won't be checked in?

Reply:
I guess if you had a binary crate in the workspace for creating the generated files, then you could build only that crate during the generate step, and generate the files.  Then they would be available to be consumed during the Rust build proper.  Would that be possible?  It would be nice to somehow make this conform with our existing build practices.

Also, I'm curious, how long does doing type inference on the standard library take?

Reply:
Those are binary files. The source files that they're derived from, i.e `builtins.rs` and the flux source files, are checked into source control though. Rust places generated files in the `target` directory when using the `OUT_DIR` environment variable, and my understanding is that the `target` directory is usually git-ignored, which it is in our case.

I am open to a better way though. This was just the first way that came to mind. Although I think we would still ignore those generated binary data files.

Reply:
Ok, none of this should stop you from merging, but the way the generated scanner works is: it lives in `libflux/scanner.c` and then `flux/build.rs` copies it to the target directory.  The makefile will build `libflux/scanner.c` during the generate step only when it gets out of sync.  

Builds in Rust are already pretty slow, so it might save some time to not infer the stdlib+prelude if we don't have to.  But like I said before, I think this work can be deferred.

Reply:
@wolffcm that makes a lot of sense. I can create an issue for this. So basically we only have to re-infer the stdlib, if one of the flux files changes or if `builtins.rs` changes. Is that what you're thinking?

Reply:
Yes, exactly.  The rule in the makefile would look something like
```makefile
stdlib.data prelude.data fresher.data: $FLUX_FILES builtins.rs
        <command to infer stdlib and prelude>
```

Reply:
@wolffcm What about if the implementation of algo w itself changes, i.e. we fix a bug? Do we need the makefile to understand all of the code that is part of the `flux` crate?

Reply:
The `checkgenerate` target should catch any issues there, same as if we fixed a parser bug and the resulting `_gen.go` changes.